sw/device/lib/base/memory_hardened.c

```diff
@@ -0,0 +1,670 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory_hardened.h"
+
+#include "sw/device/lib/base/hardened.h"
+
+static inline hardened_bool_t true_if_zero(uintptr_t val) {
+  /**
+   * We need a form where skipping instruction will not allow to
+   * produce kHardenedBoolTrue, on RV32 try to get 'snez' instead of 'seqz'
+   * as snez will transform non-zero reg into non-zero, so skipping it
+   * won't cause an inversion of value.
+   * Constant 7 is choosen to prevent input values in 0xfffffff0 - 0xffffffff
+   * range to be converted proper constant by skipping snez & neg.
+   */
+  uintptr_t mask = -(val * 7 != 0); /* val=0 -> 0, val!=0 => 0xfffffff */
+  return (hardened_bool_t)((mask & (uintptr_t)kHardenedBoolFalse) |
+                           ((~mask) & (uintptr_t)kHardenedBoolTrue));
+}
+
+/**
+ * Define machine word and alignment
+ */
+typedef uintptr_t word_t;
+#define WORD_SIZE sizeof(word_t)
+#define WORD_MASK (WORD_SIZE - 1)
+
+static inline size_t min(size_t a, size_t b) { return (a < b) ? a : b; }
+
+/**
+ * The 32 bit LFSR whose maximum length feedback polynomial is represented
+ * as X^32 + X^22 + X^2 + X^1 + 1 will produce 2^32-1 PN sequence.
+ * This LFSR can be initialized with 0,  but can't be initialized with
+ * 0xFFFFFFFF
+ */
+static uint32_t lfsr32(uint32_t seed) {
```

Comment:
Perhaps we could reuse the LFSR implementation here:
https://github.com/lowRISC/opentitan/blob/d303e90a0b7fb8f8e381e1967934da18e24e1b2d/sw/device/lib/testing/random.c#L19

Reply:
good point. I'm still learning opentitan codebase. if there is a low-entropy random number generator available not only for testing I can drop passing randomization seed with every call. on the other hand randomization seed is useful for testing.

Reply:
should I expect random_gen32() is always available? i'm concerned with it's location 'testing'.  Looking at this implementation i wonder if it's possible to use lfsr which accepts 0 as initial value, so you don't use .data section and avoid copy of initial value in RAM?

Reply:
I am inclined to leave this here for now as we need to integrate this with an entropy source. We can update this in a follow up PR.

Reply:
i'll continue to use internal implementation for a time being until better randoms are available. my implementation can be initialized with zero ('bad' value is 0xffffffff) and results in branchless code.